ui/dashboards/src/context/DatasourceStoreProvider.tsx

```diff
@@ -130,19 +131,75 @@ export function DatasourceStoreProvider(props: DatasourceStoreProviderProps) {
   );
 
   // Given a Datasource selector, finds the spec for it and then uses its corresponding plugin the create a client
-  const getDatasourceClient = useCallback(
-    async function getClient<Client extends DatasourceClient>(selector: DatasourceSelector): Promise<Client> {
+  const getSavedDatasourceClient = useCallback(
+    async (selector: DatasourceSelector) => {
       const { kind } = selector;
       const [{ spec, proxyUrl }, plugin] = await Promise.all([findDatasource(selector), getPlugin('Datasource', kind)]);
 
-      // allows extending client
-      const client = plugin.createClient(spec.plugin.spec, { proxyUrl }) as Client;
+      return plugin.createClient(spec.plugin.spec, { proxyUrl, fetch });
```

Comment:
Just skimming the PR, not crazy about passing fetch this way, but agreed this is tricky 🤔 

Is the custom version only needed for healthCheck? If so, I wonder if a similar approach to the onCreate prop could be used, so inside getDatasourceClient if the datasource is unsaved, create a wrapper that defines the custom fetch behavior and return something like: `{ …client, healthCheck: () => customFetchWrapperExample(client.healthCheck, spec) }`

Reply:
I _was_ thinking that we could apply it to things like datasources in dashboard that haven't been saved yet, but honestly if limiting it to just healthchecks makes it cleaner then I'm all for it

Reply:
Oh that is neat, sorry missed that recording before. I’ll wait for others to weigh in about expected behavior though (also curious what Grafana supports for comparison). If a particular endpoint is misconfigured, does the error message give info about which one?

Wrapper idea probably isn’t relevant if more than healthchecks though—and if it was just healthchecks, it might be possible to just tweak the logic directly in DatasourceEditorForm

Reply:
I think we should limit it to the health-checks for the moment, because creating a datasource directly in the dashboard shouldn't be the usual way. So we could restrict a bit the things you can do with local datasource